### PR TITLE
Make require case-sensitive on all platforms

### PR DESCRIPTION
**Pull request by malmaud**
Resolves https://github.com/JuliaLang/julia/issues/5789.

### CODE REVIEW

**Comment by tkelman** [comment]
I'd expect readdir to be really slow here.

**Comment by malmaud** [comment]
I don't think there's any other reliable solution. I wonder how much of a concern it is in practice - the check can be disabled on linux, which is the OS where it's most likely there will remote volumes where `readdir`'s performance could be particularly problematic. 

If the folders are local, then I'm sceptical the cost of ~20 readdir's on package src directories is meaningful compared the total  runtime of importing a package.

**Comment by malmaud** [comment]
@simonster suggested using `realpath`, which will work on Linux on OS X. Still not sure there is a great solution for windows - the best I could find is http://stackoverflow.com/questions/4763117/how-can-i-obtain-the-case-sensitive-path-on-windows, but that doesn't seem great since it will randomly fail if the underlying volume has short path names disabled.

**Comment by malmaud** [comment]
Simon correctly points out that the `realpath` solution won't work with symbolic links.

**Comment by simonster** [comment]
But you could still use it as long as `islink` returns false, and fall back to the slow path if `islink` returns true (shouldn't be very common).

**Comment by malmaud** [comment]
OK, the only case where the pr falls back to listing a directory is if the file is a symbolic link on OS X. I *think* I have a working solution for Windows - at least it works on my Windows 7 x64 box. 

https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/getattrlist.2.html might help with OS X.

File: base/loading.jl [hunk 1]
```diff
@@ -2,6 +2,28 @@
...
+    isfile(path) || return false
+    longpath(path) == path
+end
+
+@linux_only isfile_casesensitive(path) = isfile(path)
+
+@osx_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    islink(path) && return isfile_casesensitive_slow(path)
+    realpath(path) == path
```
Comment (simonster) [inline base/loading.jl]: Unless you can find a better solution, I think this should check `basename(realpath(path)) == basename(path)` in case `path` is not a symlink but is in a symlinked directory

Comment (malmaud) [inline base/loading.jl]: Yes.

**Comment by StefanKarpinski** [comment]
Excellent – thanks for tackling this.

**Comment by tkelman** [comment]
needs a test

**Comment by malmaud** [comment]
Good news - I've got working C code for getting the canonical capitalization in OS X. Just a matter of wrapping it in Julia now.

**Comment by malmaud** [comment]
OK, all the functionality is here. I'll clean it up and tests and then it should be g2g.

File: base/loading.jl [hunk 2]
```diff
@@ -2,6 +2,52 @@
 
 # Base.require is the implementation for the `import` statement
 
+# Generic case-sensitive version of isfile
+function isfile_casesensitive_slow(path)
```
Comment (stevengj) [inline base/loading.jl]: Seems like it would be cleaner to do
```jl
if OS_NAME == :WINDOWS
   function isfile_casesensitive(path)
       ...
   end
elseif OS_NAME == :Linux
    ...
else
   ...
end
```

File: base/loading.jl [hunk 3]
```diff
@@ -2,6 +2,52 @@
...
+
+isfile_casesensitive(path) = isfile_casesensitive_slow(path)
+
+@windows_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    longpath(path) == path
+end
+
+@linux_only isfile_casesensitive(path) = isfile(path)
+
```
Comment (stevengj) [inline base/loading.jl]: On MacOS, could we open the file to get a file descriptor, then call `[fcntl](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/fcntl.2.html)` with `F_GETPATH`, rather than walking the directory tree?

File: base/loading.jl (hunk 2 above)
Comment (malmaud) [inline base/loading.jl]: agreed

File: base/loading.jl (hunk 3 above)
Comment (simonster) [inline base/loading.jl]: @malmaud's code below already avoids walking the directory on OS X. `fcntl` with `F_GETPATH` seems like another possibility, although I'm not sure what it does if the file is a symlink, or if there are hard links.

Comment (malmaud) [inline base/loading.jl]: You can see below I ended up using `getattrlist`, not walking the directory tree. The code for `fcntl` would be cleaner than the complex `getattrlist` call, but I'm worried about various edge cases, performance issues, and race conditions that could happen if you actually obtain file descriptor.

**Comment by stevengj** [comment]
Also fixes #9007.

**Comment by stevengj** [comment]
See also how [Python 2.7 does it](https://github.com/python/cpython/blob/841a25de9d7efdf7e615def5622f7ea42ddae623/Python/import.c#L1618-L1798).

**Comment by malmaud** [comment]
Looks like it's walking the directory tree.

**Comment by stevengj** [comment]
Yes, looks like it.  Of course, they had to support OS/2 and other ancient systems, so maybe the things you are doing were not available at some point.

File: base/loading.jl [hunk 4]
```diff
@@ -2,6 +2,52 @@
...
+        buf = zeros(UInt8, buf_size)
+        ret = ccall(:getattrlist, Cint, (Cstring, Ptr{Void}, Ptr{Void}, Csize_t, Culong), path,
+          attr_list, buf, sizeof(buf), FSOPT_NOFOLLOW)
+        ret == 0 || error(Libc.errno())
+        len = unsafe_load(convert(Ptr{UInt32}, pointer(buf)), 3)
+        if (len + 12) >= buf_size
+            buf_size *= 2
+            continue
+        end
+        canon_path = bytestring(pointer(buf, 13))
```
Comment (stevengj) [inline base/loading.jl]: Couldn't you just do `canon_path = UTF8String(buf)` in order to avoid making a copy?

File: base/loading.jl [hunk 5]
```diff
@@ -2,6 +2,52 @@
...
+    isfile(path) || return false
+    buf_size = length(path) + 12
+    local canon_path
+    while true
+        buf = zeros(UInt8, buf_size)
+        ret = ccall(:getattrlist, Cint, (Cstring, Ptr{Void}, Ptr{Void}, Csize_t, Culong), path,
+          attr_list, buf, sizeof(buf), FSOPT_NOFOLLOW)
+        ret == 0 || error(Libc.errno())
+        len = unsafe_load(convert(Ptr{UInt32}, pointer(buf)), 3)
+        if (len + 12) >= buf_size
```
Comment (stevengj) [inline base/loading.jl]: I don't understand the `+ 12` here...

File: base/loading.jl [hunk 6]
```diff
@@ -2,6 +2,52 @@
...
+attr_list[COMMONATTR] = ATTR_CMN_NAME
+
+@osx_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    buf_size = length(path) + 12
+    local canon_path
+    while true
+        buf = zeros(UInt8, buf_size)
+        ret = ccall(:getattrlist, Cint, (Cstring, Ptr{Void}, Ptr{Void}, Csize_t, Culong), path,
+          attr_list, buf, sizeof(buf), FSOPT_NOFOLLOW)
```
Comment (stevengj) [inline base/loading.jl]: seems like it should be indented to one column after the `(`

File: base/loading.jl [hunk 7]
```diff
@@ -2,6 +2,52 @@
...
+const attr_list = zeros(UInt8, 24)
+attr_list[BITMAPCOUNT] = ATTR_BIT_MAP_COUNT
+attr_list[COMMONATTR] = ATTR_CMN_NAME
+
+@osx_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    buf_size = length(path) + 12
+    local canon_path
+    while true
+        buf = zeros(UInt8, buf_size)
```
Comment (stevengj) [inline base/loading.jl]: allocate `buf` outside the loop and use `resize!` in the loop?   Or does this have to be zeros?   If not, it would be clearer to use `Array(UInt8, buf_size)`

File: base/path.jl [hunk 8]
```diff
@@ -125,6 +125,23 @@ abspath(a::AbstractString, b::AbstractString...) = abspath(joinpath(a,b...))
...
+    p = UInt32((sizeof(path)>>2) + 1)
+    while true
+        buflength = p
+        buf = zeros(UInt16,buflength)
+        p = ccall((:GetLongPathNameW, "Kernel32"), stdcall,
+            UInt32, (Cwstring, Ptr{UInt16}, UInt32),
+            path, buf, buflength)
+        systemerror(:longpath, p == 0)
+        if (p < buflength)
+            resize!(buf, p+1)
```
Comment (stevengj) [inline base/path.jl]: If you are using `resize!` here, should probably initialize `buf = Array(UInt16, buflength)` outside the loop.   (It is confusing to use `zeros` if the initial contents of `buf` don't matter.)

File: base/path.jl [hunk 9]
```diff
@@ -125,6 +125,23 @@ abspath(a::AbstractString, b::AbstractString...) = abspath(joinpath(a,b...))
     end
 end
 
+@windows_only longpath(path::AbstractString) = longpath(utf16(path))
+@windows_only function longpath(path::UTF16String)
+    p = UInt32((sizeof(path)>>2) + 1)
```
Comment (stevengj) [inline base/path.jl]: Why are you dividing by 4 (`>>2`) instead of by 2 (`>>1`)?  Presumably what you want is the number of code units, including the null terminator.   Maybe just `UInt32(length(path.data))`?

Comment (malmaud) [inline base/path.jl]: I just blindly copied the existing `@windows_only realpath`'s
implementation to implement `longpath`, but I am also puzzled by that.

Comment (stevengj) [inline base/path.jl]: Possibly a bug?

File: base/loading.jl (hunk 4 above)
Comment (malmaud) [inline base/loading.jl]: The path is stored at an offset from the beginning of the buffer, so I had to use a somewhat inelegant workaround.

File: base/loading.jl (hunk 5 above)
Comment (malmaud) [inline base/loading.jl]: I expanded the comments to hopefully make it clearer.

File: base/loading.jl (hunk 7 above)
Comment (malmaud) [inline base/loading.jl]: agreed

File: base/path.jl (hunk 9 above)
Comment (malmaud) [inline base/path.jl]: I fixed `longpath`. Probably a new pr should fix the windows `realpath`.

File: base/loading.jl (hunk 4 above)
Comment (stevengj) [inline base/loading.jl]: Oh, I see.

File: base/path.jl [hunk 10]
```diff
@@ -125,6 +125,23 @@ abspath(a::AbstractString, b::AbstractString...) = abspath(joinpath(a,b...))
     end
 end
 
+@windows_only longpath(path::AbstractString) = longpath(utf16(path))
+@windows_only function longpath(path::UTF16String)
+    buf = Array(UInt16, length(path.data))
```
Comment (stevengj) [inline base/path.jl]: Can you also fix `realpath` accordingly?

**Comment by stevengj** [comment]
I'm not sure this is handling Unicode normalization properly.

Suppose you have a file named `"ô"`.   There are two equivalent ways to encode this string in Unicode, given by `normalize_string("ô", :NFC)` (= one codepoint U+00f4) and `normalize_string("ô", :NFD)` (= two codepoints U+006f and U+0302).

On MacOS, `isfile` returns true for either encoding.  Internally, [HFS+ converts all names to NFD](https://developer.apple.com/library/mac/qa/qa1235/_index.html), and presumably this is what `getattrlist` returns.  That means that `isfile_casesensitive(normalize_string("ô", :NFC))` will fail even though the file exists.   Solution: call `nfd(path)` before comparing, where `nfd(s::AbstractString) = isascii(s) ? s : normalize_string(s, :NFD)`.

On Windows, I'm not 100% sure of the situation.  The [documentation says](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx) *There is no need to perform any Unicode normalization on path and file name strings for use by the Windows file I/O API functions because the file system treats path and file names as an opaque sequence of WCHARs.*  This makes me think that `normalize_string("ô", :NFC)` and `normalize_string("ô", :NFD)` are distinct filenames on Windows, but it would be good to verify this (i.e. just try creating two files with these names in the same directory and see whether they co-exist).

File: base/path.jl (hunk 10 above)
Comment (malmaud) [inline base/path.jl]: Ya, but maybe it should be a separate PR just for organizational reasons.

**Comment by johansigfrids** [comment]
Yes, I can confirm `normalize_string("ô", :NFC`) and `normalize_string("ô", :NFD)`  are distinct files on Windows. Windows file system is Unicode (and UTF-16) mostly by convention, in practice its is just a `Vector{UInt16}` for which a best effort is made to display as UTF-16.

**Comment by malmaud** [comment]
Great catch @stevengj .

**Comment by nalimilan** [comment]
And what happens on OS X when a different filesystem from HFS+ is used? Should we even care about that scenario?

**Comment by malmaud** [comment]
You're right, that's a real concern. If a file is stored on a mounted non-hfs+ directory, has a unicode character, and isn't NFD-normalized, it would be invisible to `require`. Maybe we should stat the file to get the filesystem its mounted on and only do the utf normalization if its on hfs+.

**Comment by malmaud** [comment]
The same issue applies to linux: What if the file happens to be on a case-insensitive filesystem? Then loading.jl:9 is wrong. Although at least it's not as serious, since it would be no *more* wrong than the current behavior of `require`.

**Comment by malmaud** [comment]
Alright, I think I worked around the OS X non-HFS+ concern wrt utf normalization.

File: base/loading.jl [hunk 11]
```diff
@@ -2,6 +2,87 @@
...
+        end
+        # Hack to compensate for inability to create a string from a subarray with no allocations.
+        path_basename.data == casepreserved_basename && return true
+        normed_path = normalize_string(path, :NFD)
+        # If there is no match, it's possible that the file does exist but HFS+
+        # performed unicode normalization. See  https://developer.apple.com/library/mac/qa/qa1235/_index.html.
+        if normed_path == path
+            false
+        else
+            isfile_casesensitive(normed_path)
```
Comment (stevengj) [inline base/loading.jl]: Why not just:
```jl
isascii(path_basename) && return false
normed_basename = normalize_string(path_basename, :NFD)
return normed_basename.data == casepreserved_basename
```

File: base/path.jl (hunk 10 above)
Comment (stevengj) [inline base/path.jl]: Okay.

File: base/loading.jl (hunk 11 above)
Comment (malmaud) [inline base/loading.jl]: Indeed.

**Comment by stevengj** [comment]
LGTM once Travis is green and the commits are squashed.

File: test/loading.jl [hunk 12]
```diff
@@ -10,3 +10,17 @@ thefname = "the fname!//\\&\0\1*"
 @test include_string("Base.source_path()", thefname) == Base.source_path()
 @test basename(@__FILE__) == "loading.jl"
 @test isabspath(@__FILE__)
+
+let true_filename = "cAsEtEsT.jl", lowered_filename="casetest.jl"
```
Comment (stevengj) [inline test/loading.jl]: Maybe add a comment here:
```
# Issue #5789 and PR #13542:

**Comment by malmaud** [comment]
OK, squashed and tests are green.

**Comment by stevengj** [comment]
Thanks, @malmaud, for pushing this through!

**Comment by malmaud** [comment]
Thanks for the reviewing!

**Comment by stevengj** [comment]
Can you submit a quick PR for `realpath` before we forget?

File: base/loading.jl [hunk 13]
```diff
@@ -2,6 +2,84 @@
...
+# Cross-platform case-sensitive path canonicalization
+
+if OS_NAME ∈ (:Linux, :FreeBSD)
+    # Case-sensitive filesystems, don't have to do anything
+    isfile_casesensitive(path) = isfile(path)
+elseif OS_NAME == :Windows
+    # GetLongPathName Win32 function returns the case-preserved filename on NTFS.
+    function isfile_casesensitive(path)
+        isfile(path) || return false  # Fail fast
+        longpath(path) == path
```
Comment (stevengj) [inline base/loading.jl]: Possibly this should be `basename(longpath(path)) == basename(path)`?  Two reasons:

* On MacOS, we only check that the basename matches, so this is inconsistent.
* The intent of this was that `import Foo` would check for `Foo` in a case-sensitive way.  But this does more than that, and checks the whole path.  This means that users `LOAD_PATH` and `HOME` variables becomes case-sensitive, for example, which is unexpected on case-insensitive systems, and has led to at least one [confused user on discourse](https://discourse.julialang.org/t/windows-capitalization-of-the-home-environment-variable/200?u=stevengj).
